upf/upf_ipfix.c

```diff
@@ -0,0 +1,693 @@
+	  return 0;
+
+      /* Initialize the buffer */
+      b0 = context->buffers_per_worker[my_cpu_number] =
+	vlib_get_buffer (vm, bi0);
+
+      b0->current_data = 0;
+      b0->current_length = upf_ipfix_get_headersize ();
+      b0->flags |=
+	(VLIB_BUFFER_TOTAL_LENGTH_VALID | VNET_BUFFER_F_FLOW_REPORT);
+      vnet_buffer (b0)->sw_if_index[VLIB_RX] = 0;
+      vnet_buffer (b0)->sw_if_index[VLIB_TX] = exp->fib_index;
+      context->next_record_offset_per_worker[my_cpu_number] =
+	b0->current_length;
+    }
+
+  return b0;
+}
+
+static void
+upf_ipfix_export_entry (vlib_main_t * vm, flow_entry_t * f, flow_direction_t direction, u32 now)
+{
+  u32 my_cpu_number = vm->thread_index;
+  upf_ipfix_main_t *fm = &upf_ipfix_main;
+  vlib_buffer_t *b0;
+  upf_main_t *gtm = &upf_main;
+  upf_ipfix_protocol_context_t * context;
+  u16 offset;
+  upf_ipfix_template_t * template;
+  upf_session_t *sx;
+
+  /* FIXME: need to process IPFIX during session deletion */
+  if (f->ipfix_context_index == (u32) ~0 ||
+      pool_is_free_index (gtm->sessions, f->session_index))
+    return;
+
+  context = pool_elt_at_index (fm->contexts, f->ipfix_context_index);
+  ASSERT (!!ip46_address_is_ip4 (&f->key.ip[FT_ORIGIN]) == context->is_ip4);
+
+  offset = context->next_record_offset_per_worker[my_cpu_number];
+  template = upf_ipfix_templates + context->policy;
+  sx = pool_elt_at_index (gtm->sessions, f->session_index);
+
+  upf_debug ("export entry [%s], policy %u",
+	     context->is_ip4 ? "ip4" : "ip6",
+	     context->policy);
+  if (offset < upf_ipfix_get_headersize ())
+    offset = upf_ipfix_get_headersize ();
+
+  b0 = upf_ipfix_get_buffer (vm, context);
+  /* No available buffer, what to do... */
+  if (b0 == 0)
+    {
+      upf_debug ("no buffer");
+      return;
+    }
+
+  if (context->is_ip4)
+    offset += template->add_ip4_values (b0, f, direction, offset, sx);
+  else
+    offset += template->add_ip6_values (b0, f, direction, offset, sx);
+
+  /* Reset per flow-export counters */
+  flow_last_exported(f, direction) = now;
+
+  b0->current_length = offset;
+
+  context->next_record_offset_per_worker[my_cpu_number] = offset;
+  /* Time to flush the buffer? */
+  /* TODO uncomment! also: force upon removal */
+  /* if (offset + context->rec_size > exp->path_mtu) */
+  upf_ipfix_export_send (vm, b0, context, now);
+}
+
+#define vec_neg_search(v,E)			\
+  ({						\
+    word _v(i) = 0;				\
+    while (_v(i) < vec_len(v) && v[_v(i)] == E)	\
+      {						\
+	_v(i)++;				\
+      }						\
+    if (_v(i) == vec_len(v))			\
+      _v(i) = ~0;				\
+    _v(i);					\
+  })
+
+void upf_ipfix_flow_update_hook (flow_entry_t * f, flow_direction_t direction, u32 now)
+{
+  upf_ipfix_main_t *fm = &upf_ipfix_main;
+  vlib_main_t *vm = fm->vlib_main;
+
+  if (fm->disabled || f->ipfix_context_index == (u32) ~0)
+      return;
+
+  if (fm->active_timer == 0
+      || (now > flow_last_exported(f, direction) + fm->active_timer))
```

Comment:
Thinking about NAT problem, this might be the correct place to check if `flow_last_exported` is positive before actual exporting. wdyt @ivan4th ?

Reply:
Well, in this case, it would never be exported till the flow ends ;) Added setting `flow_last_exported` to `now` upon  initialization of a particular flow